vars/artifactSetVersion.groovy

```diff
@@ -0,0 +1,138 @@
+        ]
+        Map pipelineDataMap = [
+            gitCommitId: gitUtils.getGitCommitId()
+        ]
+        List stepConfigurationKeys = [
+            'artifactType',
+            'buildTool',
+            'dockerVersionSource',
+            'filePath',
+            'gitCredentialsId',
+            'gitUserEMail',
+            'gitUserName',
+            'gitSshUrl',
+            'tagPrefix',
+            'timestamp',
+            'timestampTemplate',
+            'versioningTemplate'
+        ]
+
+        Map configuration = ConfigurationMerger.mergeWithPipelineData(parameters, parameterKeys, pipelineDataMap, stepConfiguration, stepConfigurationKeys, stepDefaults)
+
+        //buildTool is mandatory parameter
+        def utils = new Utils()
+        def buildTool = utils.getMandatoryParameter(configuration, 'buildTool')
+
+        if (!configuration.filePath)
+            configuration.filePath = configuration[buildTool].filePath //use default configuration
+
+        def newVersion
+        def artifactVersioning = ArtifactVersioning.getArtifactVersioning(buildTool, this, configuration)
+
+        if(configuration.artifactType == 'appContainer' && configuration.dockerVersionSource == 'appVersion'){
+            if (script.commonPipelineEnvironment.getArtifactVersion())
+                //replace + sign if available since + is not allowed in a Docker tag
+                newVersion = script.commonPipelineEnvironment.getArtifactVersion().replace('+', '_')
+            else
+                error ("[${stepName}] No artifact version available for 'dockerVersionSource: appVersion' -> executeBuild needs to run for the application artifact first to set the artifactVersion for the application artifact.'")
+        } else {
+            def currentVersion = artifactVersioning.getVersion()
+
+            def timestamp = configuration.timestamp ? configuration.timestamp : getTimestamp(configuration.timestampTemplate)
+
+            def versioningTemplate = configuration.versioningTemplate ? configuration.versioningTemplate : configuration[configuration.buildTool].versioningTemplate
+            //defined in default configuration
+            def binding = [version: currentVersion, timestamp: timestamp, commitId: configuration.gitCommitId]
+            def templatingEngine = new SimpleTemplateEngine()
+            def template = templatingEngine.createTemplate(versioningTemplate).make(binding)
+            newVersion = template.toString()
+        }
+
+        artifactVersioning.setVersion(newVersion)
+
+        sh 'git add .'
+
+        def gitCommitId
+
+        sshagent([configuration.gitCredentialsId]) {
+            if (configuration.gitUserName  && configuration.gitUserEMail) {
+                sh "git config user.name \"${configuration.gitUserName}\""
+                sh "git config user.email \"${configuration.gitUserEMail}\""
+            }
+            try {
+                sh "git commit -m 'update version ${newVersion}'"
```

Comment:
In which cases  do we put something into GitUtils and in which cases there is a direct git invocation? Or with other words: wouln't it make sense to provide GitUtils#commit().

Refactoring would be easier in case we have to e.g. exchange the way we interact with git in the future.

Applies also for other git invocations here in this file

Reply:
I so far foresee git commit and push only here. I think only when we have reuse cases we should put it into the GitUtils. 
The reuse is clearly there for retrieving the current commitId.
What do you think?

Reply:
I'm not blessed with a doctrinal/orthodox mind set, it is about having code with can be maintained in an efficient manner later on.

That maintainablity aspect is the other one beside the re-use aspect. It is easier to e.g. replace the way we interact with git (eg. switch from sh based calls to egit (I do not endorse such a switch, it is just an example)). If we have a git util class anyway I would expect something rather in that util class than here.

But in case you are not convinced, keep it as it is.

Reply:
I would be with you but since we have only a few occurences in one step so far I think it is fine to keep it as is since also using another git implementation other than the one on OS is pretty unlikely.